LANG-1504 - Adding steps feature to StopWatch

topera opened the pull request.

garydgregory commented:
FYI, the conventional name for what you call "step" is "split", sometimes also "lap" (less modern terminology.)

topera commented:
> FYI, the conventional name for what you call "step" is "split", sometimes also "lap" (less modern terminology.)

I didn`t know that, thanks for sharing :)

I take a look now and the StopWatch already has the **split** method, but works in a different way.

Some differences:
1. With **split** we need to call **getSplitTime** while we are in the middle of split (I mean, before calling **unsplit**). In the "steps" solution we can get all times in a separated report, at the end of our performance test, for example.

2. With **split** we can have just one split each time. With **steps**, many

3. With **steps** we can label each step with a name

4. Split counts the time since the StopWatch started, not the time between splits.
Example

       ```
 final StopWatch sw = StopWatch.createStarted();

        Thread.sleep(100);

        sw.split();

        Thread.sleep(200);

        sw.split();

        Thread.sleep(400);

        System.out.println("splitTime: " + sw.getSplitTime()); // returns 300: the time between start and latest split
        System.out.println("time: " + sw.getTime()); // returns 700

```

So, my proposal is to use a new name no change the idea of the old split.
Possibilities: **lap** (as you said), **step**, **stage**

garydgregory commented:
Since we already have split() we should keep the name. Another idea, is
that split times should be kept in history list. But then, I'd want a list
of pairs (message-split time) and it starts to feel like feature creep ;-)

I'm not sure how handy this would as I do not have a use case for it.

I'd like to cut a RC later this week. There is already plenty for a release
ATM but let's keep the discussion going :-)

Gary


On Mon, Dec 30, 2019 at 7:22 PM Rafael Topera <notifications@github.com>
wrote:

> FYI, the conventional name for what you call "step" is "split", sometimes
> also "lap" (less modern terminology.)
>
> I didn`t know that, thanks for sharing :)
>
> I take a look now and the StopWatch already has the *split* method, but
> works in a different way.
>
> Some differences:
>
>    1.
>
>    With *split* we need to call *getSplitTime* while we are in the middle
>    of split (I mean, before calling *unsplit*). In the "steps" solution
>    we can get all times in a separated report, at the end of our performance
>    test, for example.
>    2.
>
>    With *split* we can have just one split each time. With *steps*, many
>    3.
>
>    With *steps* we can label each step with a name
>    4.
>
>    Split counts the time since the StopWatch started, not the time
>    between splits.
>    Example
>
>    ```
>
>
> final StopWatch sw = StopWatch.createStarted();
>
>     Thread.sleep(100);
>
>     sw.split();
>
>     Thread.sleep(200);
>
>     sw.split();
>
>     Thread.sleep(400);
>
>     System.out.println("splitTime: " + sw.getSplitTime()); // returns 300: the time between start and latest split
>     System.out.println("time: " + sw.getTime()); // returns 700
>
>
> So, my proposal is to use a new name no change the idea of the old split.
> Possibilities: **lap** (as you said), **step**, **stage**
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-lang/pull/482?email_source=notifications&email_token=AAJB6N2KGGVNWSYE637EDE3Q3KGE3A5CNFSM4KBARU2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH3P46Y#issuecomment-569835131>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJB6NYOU656F74UM24RTRLQ3KGE3ANCNFSM4KBARU2A>
> .
>

topera commented:
Hello Mr. Gregory,

Thank you for your answer.
I'm also glad that part of my suggestion was already accepted and done on LANG-1506 (putting labels on the timer)

About your points:
* I agree to use **split** name.
* About **feature creep**, IMHO this is a good feature: with small lines of code we can make the StopWatch much more useful, mainly to detect performance bottlenecks on the code.
* Use case: the developer wants to know which operation is slower. He/she can simply add many split() calls and label then as follow:

```
        final StopWatch watch = StopWatch.createStarted();

        watch.split("Baking cookies");
        ...
        watch.split("Serving cookies");
        ...
        watch.split("Eating cookies");
        ...
        watch.getSplitReport();

```
The output would be (with or without the percentage indicator)

```
        Baking cookies  00:07:00.000 - 70%
        Serving         00:01:00.000 - 10%
        Eating          00:02:00.000 - 20%
```

If you and other members of the team agrees, I'll proceed with the implementation.
Note: this is my first try on open source projects, so please sorry if I'm skipping some important step on the process.

Regards

kinow commented:
Haven't had time to review this change @topera. But

>Note: this is my first try on open source projects, so please sorry if I'm skipping some important step on the process.

you are doing great, being patient, and trying to express your point :+1: thanks for that.

It would be great if you fixed the conflicting files in this PR. There are two files in conflict, see screenshot:

![image](https://user-images.githubusercontent.com/304786/72674370-63363580-3ada-11ea-8dd2-b324438049ba.png)

garydgregory commented:
Let's have this PR's conflicts resolved, rename "steps" to "splits" to match the current terminology and then we can have another look.

topera commented:
PR updated according to suggestions

garydgregory commented on src/main/java/org/apache/commons/lang3/time/StopWatch.java:
```diff
@@ -523,10 +586,17 @@ public void stop() {
         }
         if (this.runningState == State.RUNNING) {
             this.stopTime = System.nanoTime();
+            split("");
```
Please reuse StringUtils.EMPTY.

topera replied:
Done

garydgregory commented:
Thanks for the update. I will review later this week but consider java.time classes for more precision.

garydgregory commented:
As you can see here "All checks have failed", please fix those up. 
https://travis-ci.org/apache/commons-lang/builds/654709829?utm_source=github_status&utm_medium=notification

topera commented on src/main/java/org/apache/commons/lang3/time/StopWatch.java:
```diff
@@ -606,4 +681,175 @@ public void unsplit() {
+     * @param useNanos if must use nanoseconds precision
+     * @return list of splits
+     */
+    public List<Split> getProcessedSplits(boolean useNanos) {
```
Good idea, since it makes the code more clear. I'll change here. Thank you!
